tools/performance-server/src/main/kotlin/database/BenchmarksIndexesDispatcher.kt

```diff
@@ -0,0 +1,323 @@
+                                        "filters": { "samples": { "match": { "benchmarks.status": "FAILED" } } }
+                                    },
+                                    "aggs" : {
+                                        "failed_count": {
+                                            "value_count": {
+                                                "field" : "benchmarks.score"
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                    ${buildNumbers?.let {
+            """ }
+                }"""
+        } ?: ""}
+        }
+    }
+}
+"""
+        return getIndex(featureValue).search(queryDescription, listOf("aggregations")).then { responseString ->
+            val dbResponse = JsonTreeParser.parse(responseString).jsonObject
+            val aggregations = dbResponse.getObjectOrNull("aggregations") ?: error("Wrong response:\n$responseString")
+            buildNumbers?.let {
+                // Get failed number for each provided build.
+                val buckets = aggregations.getObjectOrNull("builds")?.getObjectOrNull("buckets")
+                        ?: error("Wrong response:\n$responseString")
+                buildNumbers.map {
+                    it to buckets.getObject(it).getObject("metric_build").getObject("metric_samples")
+                            .getObject("buckets").getObject("samples").getObject("failed_count").getPrimitive("value").int
+                }.toMap()
+            } ?: listOf("golden" to aggregations.getObject("metric_build").getObject("metric_samples")
+                    .getObject("buckets").getObject("samples").getObject("failed_count").getPrimitive("value").int).toMap()
+        }
+    }
+
+    // Get geometric mean for benchmarks values of needed metric.
+    fun getGeometricMean(metricName: String, featureValue: String = "",
+                         buildNumbers: Iterable<String>? = null, normalize: Boolean = false,
+                         excludeNames: List<String> = emptyList()): Promise<List<Pair<String, List<Double>>>> {
+        // Filter only with metric or also with names.
+        val filterBenchmarks = if (excludeNames.isEmpty())
+            """
+            "match": { "benchmarks.metric": "$metricName" }
+            """
+        else """
+            "bool": { 
+                "must": { "match": { "benchmarks.metric": "$metricName" } },
+                "must_not": { "terms" : { "benchmarks.name" : [${excludeNames.map { "\"$it\"" }.joinToString()}] } }
+            }
+        """.trimIndent()
+        val queryDescription = """
+            {
+                "_source": false,
+                ${featureFilter?.let {
+            """
+                "query": {
+                    "bool": {
+                        "must": [ ${it(featureValue)} ]
+                    }
+                }, """
+        } ?: ""}
```

Comment:
I think making your own function like:
```
inline fun <T> T.str(block: (T) -> String): String =
   if (this != null) block(this)
   else ""
```  
may improve readability and remove cluttering with several `let` and `?: ""`